sycl/test-e2e/FilterSelector/filter_list_cpu_gpu_acc.cpp

```diff
@@ -6,7 +6,7 @@
 //
 //===---------------------------------------------------------------------===//
 
-// REQUIRES: cpu, gpu, accelerator
+// REQUIRES: any-device-is-cpu, any-device-is-gpu, any-device-is-accelerator
```

Comment:
Ok, at this point I've probably realized how it works. `any-device-is-*` is computed not for a _current_ device that we are running on, but for the whole environment we are in.

Therefore, in an environment where we have all three types of devices, all three `any-device -is*` will be evaluated to `true`.

Even though I appreciate the change, I don't think that it should be accepted without update to the documentation. Otherwise, someone could introduce a new incorrect test literally tomorrow.

Moreover, I would maybe think about alternative solutions here. Like, my first though is: why `gpu` doesn't just work as `any-device-is-gpu`? And the answer here is that sometimes we want a test to only be run **on** GPU and not just because GPU is present on a system.

Would it be better to rename `any-device-is-cpu` to something like `cpu-device-is-present`? To clarify that we don't require a test to be launched on CPU, but we require a presence of a device on the system. Or is it only me who got confused here?

Reply:
Why not just use C expression which is intuitive to everyone?
```
linux && cpu   # linux and cpu env
linux && gpu  # linux gpu env
linux && (gpu || cpu) # linux, cpu env or gpu env
linux && (cpu && gpu)  # linux, cpu & gpu must exist at the same time
linux && !acc  # linux, cpu env or gpu env or cpu&gpu env
```

Reply:
We need to get rid of "," expression

Reply:
> Even though I appreciate the change, I don't think that it should be accepted without update to the documentation. Otherwise, someone could introduce a new incorrect test literally tomorrow.

100% agree with the documentation change given how easy it is to make this error. Not to mention that it doesn't just affect `REQUIRES`, but also any lit line that doesn't have a `%{run}`. Any statement like `%if <device/backend>` never evaluates to true unless there is a `%{run}` in the line. For example:
https://github.com/intel/llvm/blob/sycl/sycl/test-e2e/Basic/interop/interop_all_backends.cpp#L2.

> Why not just use C expression which is intuitive to everyone?
> 
> ```
> linux && cpu   # linux and cpu env
> linux && gpu  # linux gpu env
> linux && (gpu || cpu) # linux, cpu env or gpu env
> linux && (cpu && gpu)  # linux, cpu & gpu must exist at the same time
> linux && !acc  # linux, cpu env or gpu env or cpu&gpu env
> ```
This would still have the issue that this PR is trying to fix, how we currently handle requires `linux && (cpu && gpu)` would never be met. However if we got the conditions to work as described, i feel like this would be quite intuitive.

> Moreover, I would maybe think about alternative solutions here. Like, my first though is: why `gpu` doesn't just work as `any-device-is-gpu`? And the answer here is that sometimes we want a test to only be run **on** GPU and not just because GPU is present on a system.

Honestly, I like this solution, we could structure the tests slightly differently in the two cases you mention. If the intention is to run because a GPU is present in the system we would likely also add another device to the requires, and include logic (either through device selectors, or env variables) to make sure we are running on the device we intend. In the case we want to specifically only run on a GPU we would use `%{run}` which would take care of that.

Reply:
> This would still have the issue that this PR is trying to fix, how we currently handle requires linux && (cpu && gpu) would never be met.

I see. Some of my tests require both cpu and gpu device available, but it's fine to ignore these tests if we don't have this environment currently.

Reply:
> > This would still have the issue that this PR is trying to fix, how we currently handle requires linux && (cpu && gpu) would never be met.
> 
> I see. Some of my tests require both cpu and gpu device available, but it's fine to ignore these tests if we don't have this environment currently.

@AllanZyne, the problem is confusion between "it required to be run on GPU" vs "it requires a GPU to be present on the system, but can be launched on other device". When you write `REQUIRES: cpu, gpu` you probably mean that the test requires both to be present on the system, but can be launched on either of them. But how it is perceived by the testing framework is that "test must be run on a device which is both CPU & GPU at the same time" which is never true.

To me, `any-device-is-cpu` is a bit unclear alias to "cpu device is present in the environment" and I don't think that we will be able to make `cpu` mean `any-device-is-cpu`, because we do have cases where we specifically want to run a test only on `cpu`. Therefore, as a relatively cheap trade-off, we could consider renaming `any-device-is-cpu` to `cpu-is-present`, for example.

Reply:
> Therefore, as a relatively cheap trade-off, we could consider renaming `any-device-is-cpu` to `cpu-is-present`, for example.

Should we also rename `cpu` to `runs-on-cpu`? Then everything will be self-documenting

Reply:
I'm in favor of `cpu-is-present`, but I don't like the `runs-on-cpu`. IMO, it's even more ambiguous than before.

Reply:
if the true intent of "REQUIRES: cpu,gpu,accelerator" is tested on any of "cpu, gpu, acc", why not just rewrite it to "REQUIRES: cpu || gpu || accelerator"?

Reply:
> @AllanZyne, the problem is confusion between "it required to be run on GPU" vs "it requires a GPU to be present on the system, but can be launched on other device". 

Do we really need "it requires a GPU to be present on the system, but can be launched on other device" this case?
If so, I think it's more intuitive to add a new label, such as

```
REQUIRE: cpu
MACHINE: cpu && gpu  // or, "cpu, gpu"
```

It means this test needs to run one a machine which has cpu and gpu, but this test will only run on cpu. Here, the "REQUIRE" means which devices it is going to test on.

> When you write REQUIRES: cpu, gpu you probably mean that the test requires both to be present on the system, but can be launched on either of them.

```
REQUIRE: cpu || gpu
MACHINE: cpu && gpu  // or, "cpu, gpu"
```

And by default,

```
REQUIRE: cpu || gpu
```

It means this test can be tested on cpu or gpu, it doesn't require the machine have both cpu and gpu. If the machine only have cpu, it's fine to just test cpu. If the machine have cpu and gpu, it will test on cpu and gpu separately (run twice).

```
REQUIRE: cpu && gpu
```

It means, it need a machine have both cpu and gpu, and this test need these device both available (run once).

Reply:
I means "REQUIRE" is used to control "ONEAPI_DEVICE_SELECTOR",

```
REQUIRE: cpu || gpu
```

export ONEAPI_DEVICE_SELECTOR=opencl:cpu
export ONEAPI_DEVICE_SELECTOR=levezero:gpu

```
REQUIRE: cpu && gpu
```

export ONEAPI_DEVICE_SELECTOR=levezero:gpu,opencl:cpu